core/modules/parser/QSMySqlListener.cc

```diff
@@ -2725,47 +2717,74 @@ class LogicalExpressionAdapter :
     }
 
     void onExit() override {
-        ASSERT_EXECUTION_CONDITION(_logicalOperator != nullptr, "logicalOperator is not set.", _ctx);
-
-        bool isOr = dynamic_pointer_cast<query::OrTerm>(_logicalOperator) != nullptr;
-        for (auto term : _terms) {
-            if (false == _logicalOperator->merge(*term)) {
-                if (isOr) {
-                    _logicalOperator->addBoolTerm(make_shared<query::AndTerm>(term));
-                } else {
-                    _logicalOperator->addBoolTerm(term);
+        ASSERT_EXECUTION_CONDITION(_logicalOperatorIsSet, "logicalOperator is not set.", _ctx);
+        shared_ptr<query::LogicalTerm> logicalTerm;
+        switch (_logicalOperatorType) {
+            case LogicalOperatorCBH::AND: {
+                logicalTerm = make_shared<query::AndTerm>();
+                for (auto term : _terms) {
+                    if (false == logicalTerm->merge(*term)) {
+                        logicalTerm->addBoolTerm(term);
+                    }
+                }
+                break;
+            }
+
+            case LogicalOperatorCBH::OR: {
+                logicalTerm = make_shared<query::OrTerm>();
+                for (auto term : _terms) {
+                    if (false == logicalTerm->merge(*term)) {
+                        logicalTerm->addBoolTerm(make_shared<query::AndTerm>(term));
+                    }
+                }
+                break;
+            }
+
+            case LogicalOperatorCBH::XOR: {
+                logicalTerm = make_shared<query::OrTerm>();
+                for (auto addTerm : _terms) {
+                    query::BoolTerm::PtrVector terms;
+                    for (auto otherTerm: _terms) {
+                        if (otherTerm == addTerm) {
+                            terms.push_back(otherTerm);
+                        } else {
+                            auto term = dynamic_pointer_cast<query::BoolFactor>(otherTerm->copySyntax());
+                            ASSERT_EXECUTION_CONDITION(term != nullptr, "XOR currently only works with BoolFactor", _ctx);
+                            term->setHasNot(true);
```

Comment:
Does this handle terms that already have NOT, I mean something like:

    A XOR NOT B